File: packages/api/api/concepts/stellar_cartography/black_holes.py
"""Black hole ergosphere geometry aligned with Planets.nu client ``getBlackHoleBand``."""

from __future__ import annotations

import math

# Planets.nu host constants; contract: test-fixtures/black-hole-ergosphere-contract.json.
ERGOSPHERE_BAND_COUNT = 9
BLACK_HOLE_HALO_EXTRA_LY = 5


def ergosphere_outer_radius(coreradius: int, bandradius: int) -> int:
    """Outer edge of the ergosphere in ly from the black hole center."""
    return coreradius + ERGOSPHERE_BAND_COUNT * bandradius


def black_hole_band_at(coreradius: int, bandradius: int, dist: float) -> int | None:
    """Return band index at ``dist`` ly from center, or ``None`` if outside the ergosphere.

    ``0`` is the lethal core. Bands ``1``..``9`` run inward-to-outward (1 = innermost).
    """
    if bandradius <= 0:
        return None
    outer = ergosphere_outer_radius(coreradius, bandradius)
    if dist > outer:
        return None
    if dist <= coreradius:
        return 0
    return min(
        ERGOSPHERE_BAND_COUNT,
        max(1, math.ceil((dist - coreradius) / bandradius)),
    )


def black_hole_max_warp_at(coreradius: int, bandradius: int, dist: float) -> int | None:
    """Max safe ordered warp at ``dist`` ly; ``None`` in core or outside ergosphere."""
    band = black_hole_band_at(coreradius, bandradius, dist)
    if band is None or band == 0:
        return None
    return band


def black_hole_fuel_saving_percent_at(coreradius: int, bandradius: int, dist: float) -> int | None:
    """Host predictor fuel bonus percent at ``dist`` ly; ``None`` in core or outside."""
    band = black_hole_band_at(coreradius, bandradius, dist)
    if band is None or band == 0:
        return None
    return 10 - band

File: packages/frontend/src/lib/cartography/blackHoles.ts
/**
 * Black hole ergosphere sampling aligned with Planets.nu ``getBlackHoleBand``.
 * Duplicated in api/concepts/stellar_cartography/black_holes.py; keep in sync via
 * test-fixtures/black-hole-ergosphere-contract.json.
 */

import {
  ERGOSPHERE_BAND_COUNT,
  blackHoleErgosphereOuterLy,
  blackHoleHaloRadiusLy,
} from './stellarCartographyTheme'

export { ERGOSPHERE_BAND_COUNT, blackHoleErgosphereOuterLy, blackHoleHaloRadiusLy }

/** Band index at ``distLy`` from center, or ``null`` outside the ergosphere. Band ``0`` is lethal core. */
export function blackHoleBandAt(
  coreRadiusLy: number,
  bandWidthLy: number,
  distLy: number
): number | null {
  if (bandWidthLy <= 0) {
    return null
  }
  const outer = blackHoleErgosphereOuterLy(coreRadiusLy, bandWidthLy)
  if (distLy > outer) {
    return null
  }
  if (distLy <= coreRadiusLy) {
    return 0
  }
  return Math.min(
    ERGOSPHERE_BAND_COUNT,
    Math.max(1, Math.ceil((distLy - coreRadiusLy) / bandWidthLy))
  )
}

/** Max safe ordered warp at ``distLy``; ``null`` in core or outside ergosphere. */
export function blackHoleMaxWarpAt(
  coreRadiusLy: number,
  bandWidthLy: number,
  distLy: number
): number | null {
  const band = blackHoleBandAt(coreRadiusLy, bandWidthLy, distLy)
  if (band === null || band === 0) {
    return null
  }
  return band
}

/** Host predictor fuel bonus percent at ``distLy``; ``null`` in core or outside. */
export function blackHoleFuelSavingPercentAt(
  coreRadiusLy: number,
  bandWidthLy: number,
  distLy: number
): number | null {
  const band = blackHoleBandAt(coreRadiusLy, bandWidthLy, distLy)
  if (band === null || band === 0) {
    return null
  }
  return 10 - band
}

/** Inner and outer ergosphere band edges in ly (band 1 = innermost). */
export function blackHoleBandRadiiLy(
  coreRadiusLy: number,
  bandWidthLy: number,
  band: number
): { innerLy: number; outerLy: number } {
  return {
    innerLy: coreRadiusLy + (band - 1) * bandWidthLy,
    outerLy: coreRadiusLy + band * bandWidthLy,
  }
}
